{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "da68e5f2",
   "metadata": {},
   "source": [
    "# What is about ?\n",
    "\n",
    "Analyse cell cycle for Broad Institute collection of datasets.\n",
    "\n",
    "(Ñ) A.Chervov, A. Zinovyev\n",
    "\n",
    "See paper: https://arxiv.org/abs/2208.05229 for cell cycle analysis.\n",
    "\n",
    "    \"Computational challenges of cell cycle analysis using single cell transcriptomics\"\n",
    "    Alexander Chervov, Andrei Zinovyev\n",
    "\n",
    "Data source:\n",
    "\n",
    "GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157220\n",
    "\n",
    "BROAD: https://singlecell.broadinstitute.org/single_cell/study/SCP542/pan-cancer-cell-line-heterogeneity\n",
    "\n",
    "Paper: Pan-cancer single cell RNA-seq uncovers recurring programs of cellular heterogeneity https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8135089/\n",
    "\n",
    "Gabriela S. Kinker,,1,4 Alissa C. Greenwald,,1 Rotem Tal,1 Zhanna Orlova,1 Michael S. Cuoco,2 James M. McFarland,3 Allison Warren,3 Christopher Rodman,2 Jennifer A. Roth,3 Samantha A. Bender,3 Bhavna Kumar,5 James W. Rocco,5 Pedro ACM Fernandes,4 Christopher C. Mader,3 Hadas Keren-Shaul,6,7 Alexander Plotnikov,6 Haim Barr,6 Aviad Tsherniak,3 Orit Rozenblatt-Rosen,2 Valery Krizhanovsky,1 Sidharth V. Puram,8 Aviv Regev,2 and Itay Tirosh1,#\n",
    "\n",
    "\n",
    "\n",
    "### Versions:\n",
    "\n",
    "7 - cosmetic changes \n",
    "\n",
    "6 - added PCA plots with more axis 1,3 and 2,3 from Tirosh genes\n",
    "    \n",
    "    # not see something interesting \n",
    "\n",
    "5 - added PCA plots from reactome-ext list and from ALL genes\n",
    "    \n",
    "    # Results are not very clear - expected: reactome is the same as Tirosh (plus minus)\n",
    "    # In fact sometimes it shows \"circle+hole\" worse, sometimes better , many times - more or less the same \n",
    "    \n",
    "\n",
    "4 - added PCA plot from Tirosh genes\n",
    "\n",
    "    # Only for 3 examples of the fast cycle, PCA shows cycle clearly (hole is seen) \n",
    "    # Actually these examples are kind of 4-angle, and 2 more examples 4-angle - that is nice - should be investigated further\n",
    "    \n",
    "3 - same as 2 minor corrections \n",
    "\n",
    "2 - phase plots for all processed datasets \n",
    "\n",
    "1 - just phase plot for one particular dataset  - no clear picture of the cell cycle \n",
    "\n",
    "    # path_and_filename = '/kaggle/input/scrnaseq-collection-of-cancer-cell-lines/processed_h5ad/HEC59_ENDOMETRIUM_proc.h5ad'\n",
    "    # str_data_inf = 'HEC59 ENDOMETRIUM'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4b749dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load\n",
    "\n",
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "\n",
    "# Input data files are available in the read-only \"../input/\" directory\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n",
    "\n",
    "import os\n",
    "c = 0\n",
    "for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "    for filename in filenames:\n",
    "        c+=1\n",
    "        if c>=10: break\n",
    "        print(c, os.path.join(dirname, filename))\n",
    "            \n",
    "# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n",
    "# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fdc7f64",
   "metadata": {},
   "source": [
    "# Install/import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a680ecb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install scanpy \n",
    "import scanpy as sc\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy \n",
    "\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "t0start = time.time()\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['figure.dpi'] = 70\n",
    "#plt.style.use('dark_background')\n",
    "import seaborn as sns\n",
    "\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9444e9bd",
   "metadata": {},
   "source": [
    "# Cell cycle phase plots\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8633db9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "S_phase_genes_Tirosh = ['MCM5', 'PCNA', 'TYMS', 'FEN1', 'MCM2', 'MCM4', 'RRM1', 'UNG', 'GINS2', 'MCM6', 'CDCA7', 'DTL', 'PRIM1', 'UHRF1', 'MLF1IP', 'HELLS', 'RFC2', 'RPA2', 'NASP', 'RAD51AP1', 'GMNN', 'WDR76', 'SLBP', 'CCNE2', 'UBR7', 'POLD3', 'MSH2', 'ATAD2', 'RAD51', 'RRM2', 'CDC45', 'CDC6', 'EXO1', 'TIPIN', 'DSCC1', 'BLM', 'CASP8AP2', 'USP1', 'CLSPN', 'POLA1', 'CHAF1B', 'BRIP1', 'E2F8']\n",
    "G2_M_genes_Tirosh = ['HMGB2', 'CDK1', 'NUSAP1', 'UBE2C', 'BIRC5', 'TPX2', 'TOP2A', 'NDC80', 'CKS2', 'NUF2', 'CKS1B', 'MKI67', 'TMPO', 'CENPF', 'TACC3', 'FAM64A', 'SMC4', 'CCNB2', 'CKAP2L', 'CKAP2', 'AURKB', 'BUB1', 'KIF11', 'ANP32E', 'TUBB4B', 'GTSE1', 'KIF20B', 'HJURP', 'CDCA3', 'HN1', 'CDC20', 'TTK', 'CDC25C', 'KIF2C', 'RANGAP1', 'NCAPD2', 'DLGAP5', 'CDCA2', 'CDCA8', 'ECT2', 'KIF23', 'HMMR', 'AURKA', 'PSRC1', 'ANLN', 'LBR', 'CKAP5', 'CENPE', 'CTCF', 'NEK2', 'G2E3', 'GAS2L3', 'CBX5', 'CENPA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c556dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://raw.githubusercontent.com/chervov/genes/main/cell_cycle_genes_reactome_extended.txt'\n",
    "cell_cycle_genes_reactome_extended = pd.read_csv(url,header= None)\n",
    "cell_cycle_genes_reactome_extended = list(cell_cycle_genes_reactome_extended[0].unique())\n",
    "print( len(cell_cycle_genes_reactome_extended), cell_cycle_genes_reactome_extended[:10] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "441aaff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_x_subplots = 2 # \n",
    "n_y_subplots = 3\n",
    "\n",
    "import os\n",
    "import time\n",
    "\n",
    "t00 = time.time()\n",
    "\n",
    "verbose = 1\n",
    "cc = 0\n",
    "IX = 0\n",
    "df_stat = pd.DataFrame()\n",
    "for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "    if 'processed_h5ad' not in  dirname: continue \n",
    "        \n",
    "    for filename in filenames:\n",
    "        cc+=1\n",
    "        #if cc>=4: break\n",
    "\n",
    "        path_and_filename = os.path.join(dirname, filename)\n",
    "        str_data_inf = filename.replace('_', ' ')\n",
    "        \n",
    "        if verbose >= 1:\n",
    "            print(cc, path_and_filename )\n",
    "        \n",
    "        adata_orig = sc.read(path_and_filename)\n",
    "        # adata_orig = sc.read_10x_mtx(\n",
    "        #     path_and_filename, # 'data/filtered_gene_bc_matrices/hg19/',  # the directory with the `.mtx` file\n",
    "        #     var_names='gene_symbols',                # use gene symbols for the variable names (variables-axis index)\n",
    "        #     cache=True)  \n",
    "\n",
    "        adata = adata_orig.copy()\n",
    "        if verbose >= 1000:\n",
    "            print(adata.X.sum(axis=1).ravel()[:5] )\n",
    "            print( adata )\n",
    "            \n",
    "        \n",
    "        mask = pd.Series(index = adata.obs.index, data = True) # Just all include mask - in future may be change to something\n",
    "\n",
    "        fig = plt.figure(figsize = (25,20));  c = 0\n",
    "        #plt.suptitle(str_data_inf, fontsize = 20)       \n",
    "        cell_type = ''\n",
    "        plt.suptitle(cell_type +' '+str(mask.sum())+'Cells'  + ' ' + str_data_inf ,fontsize= 18  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "\n",
    "        if n_x_subplots*n_y_subplots >= 1: # Phase plot\n",
    "            \n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "\n",
    "            list_genes_upper = [t.upper() for t in adata.var.index ]\n",
    "            I = np.where( pd.Series(list_genes_upper).isin( S_phase_genes_Tirosh ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            v1 = adata[mask,:].X[:,I].mean(axis = 1)\n",
    "            v1 = np.asarray(v1).ravel()\n",
    "            I = np.where( pd.Series(list_genes_upper).isin( G2_M_genes_Tirosh ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            v2 = adata[mask,:].X[:,I].mean(axis = 1)\n",
    "            v2 = np.asarray(v2).ravel()\n",
    "\n",
    "            corrcoef = np.round(np.corrcoef(v1,v2)[0,1] , 2)\n",
    "            IX+=1\n",
    "            df_stat.loc[IX,'Correlation G1S vs G2M'] = corrcoef\n",
    "            df_stat.loc[IX,'n_cells'] = adata.X.shape[0];         df_stat['n_cells'] = df_stat['n_cells'].astype(int)\n",
    "            df_stat.loc[IX,'Name'] = str_data_inf\n",
    "\n",
    "\n",
    "            col =  'stage' # 'total_counts' #'pct_counts_mt' # \n",
    "            col = 'cell_type' #  'n_counts' # 'age', 'sex', 'sample', 'patient', 'cell_type'\n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "            #plt.title(cell_type +' '+str(mask.sum())+'Cells'  + ' ' + str_data_inf + ' Phase plot',fontsize= 20  )\n",
    "            cell_type = ''\n",
    "            #plt.title('Correlation G1SvsG2M='+str(corrcoef) +  ' '+cell_type +' '+str(mask.sum())+'Cells'  + ' ' + str_data_inf + ' Phase plot',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.title('Correlation G1SvsG2M='+str(corrcoef) +   ' Phase plot',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel('G1S',  fontsize = 20)\n",
    "            plt.ylabel('G2M',  fontsize = 20)\n",
    "            \n",
    "            \n",
    "        if n_x_subplots*n_y_subplots >= 2: # PCA plot from Tirosh\n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "            list_genes_upper = [t.upper() for t in adata.var.index ]\n",
    "            I = np.where( pd.Series(list_genes_upper).isin( S_phase_genes_Tirosh + G2_M_genes_Tirosh ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            \n",
    "            from sklearn.decomposition import PCA\n",
    "            reducer = PCA(n_components=2)\n",
    "            r = reducer.fit_transform(adata[mask,:].X[:,I])\n",
    "            v1,v2 = r[:,0],r[:,1]\n",
    "            \n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "\n",
    "            plt.title(str(reducer)+ ' From Tirosh genes',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel(str(reducer)+'1',  fontsize = 20)\n",
    "            plt.ylabel(str(reducer)+'2',  fontsize = 20)\n",
    "\n",
    "        if n_x_subplots*n_y_subplots >= 3: # PCA plot from Tirosh\n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "            list_genes_upper = [t.upper() for t in adata.var.index ]\n",
    "            I = np.where( pd.Series(list_genes_upper).isin( S_phase_genes_Tirosh + G2_M_genes_Tirosh ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            \n",
    "            from sklearn.decomposition import PCA\n",
    "            reducer = PCA(n_components=3)\n",
    "            r = reducer.fit_transform(adata[mask,:].X[:,I])\n",
    "            v1,v2 = r[:,0],r[:,2]\n",
    "            \n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "\n",
    "            plt.title(str(reducer)+ ' From Tirosh genes',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel(str(reducer)+'1',  fontsize = 20)\n",
    "            plt.ylabel(str(reducer)+'3',  fontsize = 20)\n",
    "\n",
    "        if n_x_subplots*n_y_subplots >= 4: # PCA plot from Tirosh\n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "            list_genes_upper = [t.upper() for t in adata.var.index ]\n",
    "            I = np.where( pd.Series(list_genes_upper).isin( S_phase_genes_Tirosh + G2_M_genes_Tirosh ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            \n",
    "            from sklearn.decomposition import PCA\n",
    "            reducer = PCA(n_components=3)\n",
    "            r = reducer.fit_transform(adata[mask,:].X[:,I])\n",
    "            v1,v2 = r[:,1],r[:,2]\n",
    "            \n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "\n",
    "            plt.title(str(reducer)+ ' From Tirosh genes',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel(str(reducer)+'2',  fontsize = 20)\n",
    "            plt.ylabel(str(reducer)+'3',  fontsize = 20)\n",
    "            \n",
    "            \n",
    "        if n_x_subplots*n_y_subplots >= 5: # PCA from Reactom Ext Genes\n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "            list_genes_upper = [t.upper() for t in adata.var.index ]\n",
    "            I = np.where( pd.Series(list_genes_upper).isin(cell_cycle_genes_reactome_extended ) )[0]\n",
    "            if verbose >= 100:\n",
    "                print(len(I), np.array(list_genes_upper)[I])\n",
    "            \n",
    "            from sklearn.decomposition import PCA\n",
    "            reducer = PCA(n_components=2)\n",
    "            r = reducer.fit_transform(adata[mask,:].X[:,I])\n",
    "            v1,v2 = r[:,0],r[:,1]\n",
    "            \n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "\n",
    "            plt.title(str(reducer)+ ' From Reactome genes',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel(str(reducer)+'1',  fontsize = 20)\n",
    "            plt.ylabel(str(reducer)+'2',  fontsize = 20)\n",
    "\n",
    "        if n_x_subplots*n_y_subplots >= 6: # PCA from ALL Genes\n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "            from sklearn.decomposition import PCA\n",
    "            reducer = PCA(n_components=2)\n",
    "            r = reducer.fit_transform(adata[mask,:].X)\n",
    "            v1,v2 = r[:,0],r[:,1]\n",
    "            \n",
    "            col = 'total_counts'\n",
    "            if col in adata.obs.columns:\n",
    "                ax = sns.scatterplot(x = v1, y = v2 , hue = adata[mask].obs[col], palette = \"viridis\" )#, palette = ['green', 'red'] )# obs['pct_counts_mt'])\n",
    "                plt.setp(ax.get_legend().get_texts(), fontsize=20) # for legend text\n",
    "                plt.setp(ax.get_legend().get_title(), fontsize=20) # for legend title\n",
    "            else:\n",
    "                ax = sns.scatterplot(x = v1, y = v2)#  , hue = color_by, palette = ['red','green'] )# obs['pct_counts_mt'])\n",
    "\n",
    "            plt.title(str(reducer)+ ' From ALL genes',fontsize= 20  ) # 'n_neighbors='+str(n_neighbors) +\n",
    "            plt.xlabel(str(reducer)+'1',  fontsize = 20)\n",
    "            plt.ylabel(str(reducer)+'2',  fontsize = 20)\n",
    "            \n",
    "            \n",
    "        plt.show()            \n",
    "print('Finished. %.1f seconds passed total'%(-t00 + time.time()) ) \n",
    "df_stat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bca14b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist( df_stat['Correlation G1S vs G2M'], bins = 25 )\n",
    "plt.title('Correlation G1S vs G2M')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "395f4308",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a176b26",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stat.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bca7ef8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stat2 = df_stat.copy()\n",
    "df_stat2['Name'] = df_stat2['Name'].apply(lambda x: x.replace('proc.h5ad',''))\n",
    "df_stat2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "467f0f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_stat2.to_csv('cell_lines_broad_institute_info.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a64a783",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist( df_stat['Correlation G1S vs G2M'], bins = 25 )\n",
    "plt.title('Correlation G1S vs G2M')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7191d90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83e04e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(filename)\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "sc.pl.pca(adata, color='CST3')\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['CST3'])#, 'PPBP'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a346d97",
   "metadata": {},
   "outputs": [],
   "source": [
    "m = np.mean((np.exp(adata.X)-1).sum(axis = 1) )\n",
    "print(np.std((np.exp(adata.X)-1).sum(axis = 1) )/m)\n",
    "print(np.std(     adata.X.sum(axis = 1) )/ np.mean( adata.X.sum(axis = 1) ))\n",
    "\n",
    "(np.exp(adata.X)-1).sum(axis = 1)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60755e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e455e000",
   "metadata": {},
   "outputs": [],
   "source": [
    "nan"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
