{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f586c80",
   "metadata": {},
   "source": [
    "# What is about ?\n",
    "\n",
    "First look on cell cycle\n",
    "\n",
    "\n",
    "(с) A.Chervov, A. Zinovyev\n",
    "\n",
    "See paper: https://arxiv.org/abs/2208.05229\n",
    "\n",
    "\"Computational challenges of cell cycle analysis using single cell transcriptomics\" Alexander Chervov, Andrei Zinovyev\n",
    "\n",
    "\n",
    "Particular data:\n",
    "\n",
    "Paper:\n",
    "Roider, T., Seufert, J., Uvarovskii, A. et al. Dissecting intratumour heterogeneity of nodal B-cell lymphomas at the transcriptional, genetic and drug-response levels. Nat Cell Biol 22, 896–906 (2020). doi: https://doi.org/10.1038/s41556-020-0532-x\n",
    "\n",
    "Data:\n",
    "https://heidata.uni-heidelberg.de/dataset.xhtml?persistentId=doi:10.11588/data/VRJUNV\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a838ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load\n",
    "\n",
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "\n",
    "# Input data files are available in the read-only \"../input/\" directory\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n",
    "\n",
    "import os\n",
    "for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "    for filename in filenames:\n",
    "        print(os.path.join(dirname, filename))\n",
    "\n",
    "# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n",
    "# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e65f2f9",
   "metadata": {},
   "source": [
    "# Install Import modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f48c339",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install scanpy\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "\n",
    "import time\n",
    "t0start = time.time()\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "#plt.style.use('dark_background')\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b7c0543",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "634a1ef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fn = '/kaggle/input/scrnaseq-b-cells-nature-immunology-2018-gse115795/FromPANGLAODB/GSM3190075_B_cells_SRA721637_SRS3416994.mat.txt'\n",
    "# str_data_inf = ' B cells GSM3190075 '\n",
    "\n",
    "# df = pd.read_csv(fn, sep = '\\t' ) # , index_col = 0)\n",
    "# df = df.T\n",
    "# df\n",
    "\n",
    "pathname = '/kaggle/input/scrnaseq-bt-cells-and-lymphomas-natcellbiol2020/tFL1/'\n",
    "str_data_inf = ' Transformed Follicular Lymphoma 01 '\n",
    "adata = sc.read_10x_mtx(pathname)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff1a1f4a",
   "metadata": {},
   "source": [
    "# Normalize and Log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3eff6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788d9200",
   "metadata": {},
   "source": [
    "# Auxilliary function - phase plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c66854d",
   "metadata": {},
   "outputs": [],
   "source": [
    "G1S_genes_Tirosh = ['MCM5', 'PCNA', 'TYMS', 'FEN1', 'MCM2', 'MCM4', 'RRM1', 'UNG', 'GINS2', 'MCM6', 'CDCA7', 'DTL', 'PRIM1', 'UHRF1', 'MLF1IP', 'HELLS', 'RFC2', 'RPA2', 'NASP', 'RAD51AP1', 'GMNN', 'WDR76', 'SLBP', 'CCNE2', 'UBR7', 'POLD3', 'MSH2', 'ATAD2', 'RAD51', 'RRM2', 'CDC45', 'CDC6', 'EXO1', 'TIPIN', 'DSCC1', 'BLM', 'CASP8AP2', 'USP1', 'CLSPN', 'POLA1', 'CHAF1B', 'BRIP1', 'E2F8']\n",
    "G2M_genes_Tirosh = ['HMGB2', 'CDK1', 'NUSAP1', 'UBE2C', 'BIRC5', 'TPX2', 'TOP2A', 'NDC80', 'CKS2', 'NUF2', 'CKS1B', 'MKI67', 'TMPO', 'CENPF', 'TACC3', 'FAM64A', 'SMC4', 'CCNB2', 'CKAP2L', 'CKAP2', 'AURKB', 'BUB1', 'KIF11', 'ANP32E', 'TUBB4B', 'GTSE1', 'KIF20B', 'HJURP', 'CDCA3', 'HN1', 'CDC20', 'TTK', 'CDC25C', 'KIF2C', 'RANGAP1', 'NCAPD2', 'DLGAP5', 'CDCA2', 'CDCA8', 'ECT2', 'KIF23', 'HMMR', 'AURKA', 'PSRC1', 'ANLN', 'LBR', 'CKAP5', 'CENPE', 'CTCF', 'NEK2', 'G2E3', 'GAS2L3', 'CBX5', 'CENPA']\n",
    "\n",
    "# G1S_genes_Tirosh = ['ENSMUSG00000005410', 'ENSMUSG00000027342', 'ENSMUSG00000025747', 'ENSMUSG00000024742', 'ENSMUSG00000002870', 'ENSMUSG00000022673', 'ENSMUSG00000030978', 'ENSMUSG00000029591', 'ENSMUSG00000031821', 'ENSMUSG00000026355', 'ENSMUSG00000055612', 'ENSMUSG00000037474', 'ENSMUSG00000025395', 'ENSMUSG00000001228', 'ENSMUSG00000031629', 'ENSMUSG00000025001', 'ENSMUSG00000023104', 'ENSMUSG00000028884', 'ENSMUSG00000028693', 'ENSMUSG00000030346', 'ENSMUSG00000006715', 'ENSMUSG00000027242', 'ENSMUSG00000004642', 'ENSMUSG00000028212', 'ENSMUSG00000041712', 'ENSMUSG00000030726', 'ENSMUSG00000024151', 'ENSMUSG00000022360', 'ENSMUSG00000027323', 'ENSMUSG00000020649', 'ENSMUSG00000000028', 'ENSMUSG00000017499', 'ENSMUSG00000039748', 'ENSMUSG00000032397', 'ENSMUSG00000022422', 'ENSMUSG00000030528', 'ENSMUSG00000028282', 'ENSMUSG00000028560', 'ENSMUSG00000042489', 'ENSMUSG00000006678', 'ENSMUSG00000022945', 'ENSMUSG00000034329', 'ENSMUSG00000046179']\n",
    "# G2M_genes_Tirosh = ['ENSMUSG00000054717', 'ENSMUSG00000019942', 'ENSMUSG00000027306', 'ENSMUSG00000001403', 'ENSMUSG00000017716', 'ENSMUSG00000027469', 'ENSMUSG00000020914', 'ENSMUSG00000024056', 'ENSMUSG00000062248', 'ENSMUSG00000026683', 'ENSMUSG00000028044', 'ENSMUSG00000031004', 'ENSMUSG00000019961', 'ENSMUSG00000026605', 'ENSMUSG00000037313', 'ENSMUSG00000020808', 'ENSMUSG00000034349', 'ENSMUSG00000032218', 'ENSMUSG00000048327', 'ENSMUSG00000037725', 'ENSMUSG00000020897', 'ENSMUSG00000027379', 'ENSMUSG00000012443', 'ENSMUSG00000015749', 'ENSMUSG00000036752', 'ENSMUSG00000022385', 'ENSMUSG00000024795', 'ENSMUSG00000044783', 'ENSMUSG00000023505', 'ENSMUSG00000020737', 'ENSMUSG00000006398', 'ENSMUSG00000038379', 'ENSMUSG00000044201', 'ENSMUSG00000028678', 'ENSMUSG00000022391', 'ENSMUSG00000038252', 'ENSMUSG00000037544', 'ENSMUSG00000048922', 'ENSMUSG00000028873', 'ENSMUSG00000027699', 'ENSMUSG00000032254', 'ENSMUSG00000020330', 'ENSMUSG00000027496', 'ENSMUSG00000068744', 'ENSMUSG00000036777', 'ENSMUSG00000004880', 'ENSMUSG00000040549', 'ENSMUSG00000045328', 'ENSMUSG00000005698', 'ENSMUSG00000026622', 'ENSMUSG00000035293', 'ENSMUSG00000074802', 'ENSMUSG00000009575', 'ENSMUSG00000029177']\n",
    "\n",
    "def phase_plot(adata, list_genes1, list_genes2, mask = None, list_color_by = [None], str_suptitle = None, \n",
    "               str_title = None,\n",
    "               xfigsize = 20, yfigsize = 10,\n",
    "               xlabel = 'G1S score', ylabel = 'G2M score',\n",
    "               n_x_subplots = 1, n_y_subplots = 1 #  #len(list_selected_cell_lines)\n",
    "              ):\n",
    "\n",
    "    if mask is None:\n",
    "        mask = np.ones( adata.X.shape[0]).astype(bool) #  \n",
    "\n",
    "    #list_color_by = ['total_counts']# , 'Cell cycle phase']\n",
    "\n",
    "    \n",
    "    total_count = 0\n",
    "    for color_by_mode in  list_color_by: # , 'pct_counts_mt']:\n",
    "        \n",
    "        if n_x_subplots * n_y_subplots <= 1:\n",
    "            fig = plt.figure(figsize = (xfigsize,yfigsize) ); c = 0\n",
    "            if str_suptitle is not None:\n",
    "                plt.suptitle(str_suptitle + ' N_cells= ' +str(mask.sum() ) + ' ' , fontsize = 20 )#' \n",
    "        else:\n",
    "            if ( total_count % (n_x_subplots * n_y_subplots) ) == 0: \n",
    "                fig = plt.figure(figsize = (xfigsize,yfigsize) ); c = 0\n",
    "                if str_suptitle is not None:\n",
    "                    plt.suptitle(str_suptitle + ' N_cells= ' +str(mask.sum() ) + ' ' , fontsize = 20 )#' \n",
    "            c+=1; fig.add_subplot(n_y_subplots,n_x_subplots,c);\n",
    "            \n",
    "        total_count += 1\n",
    "        \n",
    "        \n",
    "        color_by_field_name = color_by_mode\n",
    "        color_by = None\n",
    "        if color_by_mode is None:\n",
    "            pass\n",
    "        elif color_by_mode in adata.var.index:\n",
    "            \n",
    "            color_by = pd.Series( np.asarray(adata[mask,color_by_mode].X.sum(axis=1) ).ravel() , name = color_by_mode )\n",
    "        elif color_by_mode in adata.obs:\n",
    "            color_by = (adata[mask].obs[color_by_field_name])\n",
    "        elif 'median_binarize_' in color_by_mode:\n",
    "            color_by_field_name = color_by_mode[16:]\n",
    "            if color_by_field_name in adata.obs:\n",
    "                color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata[mask].obs[color_by_field_name]) )\n",
    "        elif 'threshold_binarize_' in color_by_mode:\n",
    "            color_by_field_name = color_by_mode.split('_')[3]\n",
    "            threshold_binarize = float( color_by_mode.split('_')[2] )\n",
    "            if color_by_field_name in adata.obs:\n",
    "                color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n",
    "        else:\n",
    "            color_by = None\n",
    "\n",
    "        ll = list_genes1\n",
    "        I = np.where(adata[mask].var.index.isin(ll) > 0 )[0] # .sum()\n",
    "        v1 = np.asarray(adata[mask].X[:,I].mean(axis = 1)).ravel()\n",
    "        #print(len(I), v1.mean(), len(v1))    \n",
    "        \n",
    "        ll = list_genes2#  G2M_genes_Tirosh\n",
    "        I = np.where(adata[mask].var.index.isin(ll) > 0 )[0] # .sum()\n",
    "        #print(len(I))\n",
    "        v2 = np.asarray(adata[mask].X[:,I].mean(axis = 1)).ravel()    \n",
    "\n",
    "\n",
    "        if str_title is not None:\n",
    "            plt.title( str_title  , fontsize = 20 )\n",
    "        if color_by is None:\n",
    "            ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n",
    "        else:\n",
    "            #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n",
    "            ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = \"rainbow\")# 'viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n",
    "                                #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n",
    "            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n",
    "            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n",
    "\n",
    "\n",
    "        #plt.title(str_inf)# str_data_inf + ' ' + str_reducer + str_genes_inf + ' ' + col)\n",
    "        plt.xlabel(xlabel , fontsize = 20)\n",
    "        plt.ylabel(ylabel, fontsize = 20 )\n",
    "    plt.show()         \n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "264fcde8",
   "metadata": {},
   "source": [
    "# Phase plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef40aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f37bef4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "str_g1s = '''ENSMUSG00000005410,\n",
    "ENSMUSG00000027342,\n",
    "ENSMUSG00000025747,\n",
    "ENSMUSG00000024742,\n",
    "ENSMUSG00000002870,\n",
    "ENSMUSG00000022673,\n",
    "ENSMUSG00000030978,\n",
    "ENSMUSG00000029591,\n",
    "ENSMUSG00000031821,\n",
    "ENSMUSG00000026355,\n",
    "ENSMUSG00000055612,\n",
    "ENSMUSG00000037474,\n",
    "ENSMUSG00000025395,\n",
    "ENSMUSG00000001228,\n",
    "ENSMUSG00000031629,\n",
    "ENSMUSG00000025001,\n",
    "ENSMUSG00000023104,\n",
    "ENSMUSG00000028884,\n",
    "ENSMUSG00000028693,\n",
    "ENSMUSG00000030346,\n",
    "ENSMUSG00000006715,\n",
    "ENSMUSG00000027242,\n",
    "ENSMUSG00000004642,\n",
    "ENSMUSG00000028212,\n",
    "ENSMUSG00000041712,\n",
    "ENSMUSG00000030726,\n",
    "ENSMUSG00000024151,\n",
    "ENSMUSG00000022360,\n",
    "ENSMUSG00000027323,\n",
    "ENSMUSG00000020649,\n",
    "ENSMUSG00000000028,\n",
    "ENSMUSG00000017499,\n",
    "ENSMUSG00000039748,\n",
    "ENSMUSG00000032397,\n",
    "ENSMUSG00000022422,\n",
    "ENSMUSG00000030528,\n",
    "ENSMUSG00000028282,\n",
    "ENSMUSG00000028560,\n",
    "ENSMUSG00000042489,\n",
    "ENSMUSG00000006678,\n",
    "ENSMUSG00000022945,\n",
    "ENSMUSG00000034329,\n",
    "ENSMUSG00000046179'''\n",
    "\n",
    "str_g2m = '''ENSMUSG00000054717\n",
    "ENSMUSG00000019942\n",
    "ENSMUSG00000027306\n",
    "ENSMUSG00000001403\n",
    "ENSMUSG00000017716\n",
    "ENSMUSG00000027469\n",
    "ENSMUSG00000020914\n",
    "ENSMUSG00000024056\n",
    "ENSMUSG00000062248\n",
    "ENSMUSG00000026683\n",
    "ENSMUSG00000028044\n",
    "ENSMUSG00000031004\n",
    "ENSMUSG00000019961\n",
    "ENSMUSG00000026605\n",
    "ENSMUSG00000037313\n",
    "ENSMUSG00000020808\n",
    "ENSMUSG00000034349\n",
    "ENSMUSG00000032218\n",
    "ENSMUSG00000048327\n",
    "ENSMUSG00000037725\n",
    "ENSMUSG00000020897\n",
    "ENSMUSG00000027379\n",
    "ENSMUSG00000012443\n",
    "ENSMUSG00000015749\n",
    "ENSMUSG00000036752\n",
    "ENSMUSG00000022385\n",
    "ENSMUSG00000024795\n",
    "ENSMUSG00000044783\n",
    "ENSMUSG00000023505\n",
    "ENSMUSG00000020737\n",
    "ENSMUSG00000006398\n",
    "ENSMUSG00000038379\n",
    "ENSMUSG00000044201\n",
    "ENSMUSG00000028678\n",
    "ENSMUSG00000022391\n",
    "ENSMUSG00000038252\n",
    "ENSMUSG00000037544\n",
    "ENSMUSG00000048922\n",
    "ENSMUSG00000028873\n",
    "ENSMUSG00000027699\n",
    "ENSMUSG00000032254\n",
    "ENSMUSG00000020330\n",
    "ENSMUSG00000027496\n",
    "ENSMUSG00000068744\n",
    "ENSMUSG00000036777\n",
    "ENSMUSG00000004880\n",
    "ENSMUSG00000040549\n",
    "ENSMUSG00000045328\n",
    "ENSMUSG00000005698\n",
    "ENSMUSG00000026622\n",
    "ENSMUSG00000035293\n",
    "ENSMUSG00000074802\n",
    "ENSMUSG00000009575\n",
    "ENSMUSG00000029177'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb002da",
   "metadata": {},
   "outputs": [],
   "source": [
    "str_g1s = '''ENSG00000100297\n",
    "ENSG00000132646\n",
    "ENSG00000176890\n",
    "ENSG00000168496\n",
    "ENSG00000073111\n",
    "ENSG00000104738\n",
    "ENSG00000167325\n",
    "ENSG00000076248\n",
    "ENSG00000131153\n",
    "ENSG00000076003\n",
    "ENSG00000144354\n",
    "ENSG00000143476\n",
    "ENSG00000198056\n",
    "ENSG00000276043\n",
    "ENSG00000151725\n",
    "ENSG00000119969\n",
    "ENSG00000049541\n",
    "ENSG00000117748\n",
    "ENSG00000132780\n",
    "ENSG00000111247\n",
    "ENSG00000112312\n",
    "ENSG00000092470\n",
    "ENSG00000163950\n",
    "ENSG00000175305\n",
    "ENSG00000012963\n",
    "ENSG00000077514\n",
    "ENSG00000095002\n",
    "ENSG00000156802\n",
    "ENSG00000051180\n",
    "ENSG00000171848\n",
    "ENSG00000093009\n",
    "ENSG00000094804\n",
    "ENSG00000174371\n",
    "ENSG00000075131\n",
    "ENSG00000136982\n",
    "ENSG00000197299\n",
    "ENSG00000118412\n",
    "ENSG00000162607\n",
    "ENSG00000092853\n",
    "ENSG00000101868\n",
    "ENSG00000159259\n",
    "ENSG00000136492\n",
    "ENSG00000129173'''\n",
    "\n",
    "str_g2m = '''ENSG00000164104\n",
    "ENSG00000170312\n",
    "ENSG00000137804\n",
    "ENSG00000175063\n",
    "ENSG00000089685\n",
    "ENSG00000088325\n",
    "ENSG00000131747\n",
    "ENSG00000080986\n",
    "ENSG00000123975\n",
    "ENSG00000143228\n",
    "ENSG00000173207\n",
    "ENSG00000148773\n",
    "ENSG00000120802\n",
    "ENSG00000117724\n",
    "ENSG00000013810\n",
    "ENSG00000129195\n",
    "ENSG00000113810\n",
    "ENSG00000157456\n",
    "ENSG00000169607\n",
    "ENSG00000136108\n",
    "ENSG00000178999\n",
    "ENSG00000169679\n",
    "ENSG00000138160\n",
    "ENSG00000143401\n",
    "ENSG00000188229\n",
    "ENSG00000075218\n",
    "ENSG00000138182\n",
    "ENSG00000123485\n",
    "ENSG00000111665\n",
    "ENSG00000189159\n",
    "ENSG00000117399\n",
    "ENSG00000112742\n",
    "ENSG00000158402\n",
    "ENSG00000142945\n",
    "ENSG00000100401\n",
    "ENSG00000010292\n",
    "ENSG00000126787\n",
    "ENSG00000184661\n",
    "ENSG00000134690\n",
    "ENSG00000114346\n",
    "ENSG00000137807\n",
    "ENSG00000072571\n",
    "ENSG00000087586\n",
    "ENSG00000134222\n",
    "ENSG00000011426\n",
    "ENSG00000143815\n",
    "ENSG00000175216\n",
    "ENSG00000138778\n",
    "ENSG00000102974\n",
    "ENSG00000117650\n",
    "ENSG00000092140\n",
    "ENSG00000139354\n",
    "ENSG00000094916\n",
    "ENSG00000115163'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b058d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "list2 = str_g2m.split('\\n')# ,'').split(',')\n",
    "print(list2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e2f668",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list1 = str_g1s.replace('\\n','').split(',')\n",
    "list1 = str_g1s.split('\\n')# ,'').split(',')\n",
    "\n",
    "print(list1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bf15457",
   "metadata": {},
   "outputs": [],
   "source": [
    "phase_plot(adata, list_genes1 =G1S_genes_Tirosh , # list1, #  \n",
    "           list_genes2= G2M_genes_Tirosh, # list2, # \n",
    "           mask = None,str_title = 'Tirosh genes' ,\n",
    "               str_suptitle = str_data_inf + ' n_cells = ' + str( adata.shape[0] ),\n",
    "          list_color_by = ['Cell Type',   'CCNB1','CCNE1'],)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eb555fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f3b2366",
   "metadata": {},
   "outputs": [],
   "source": [
    "nan"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
